src/com/google/enterprise/adaptor/DocRequest.java

```diff
@@ -0,0 +1,113 @@
+// Copyright 2011 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.enterprise.adaptor;
+
+import com.google.enterprise.adaptor.testing.UnsupportedRequest;
+
+import java.util.Date;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A complete implementation of {@link Request}.
+ *
+ * @see UnsupportedRequest
+ */
+public class DocRequest implements Request {
+  private static final Logger log
+      = Logger.getLogger(DocRequest.class.getName());
+
+  private final DocId docId;
+  private final Date lastAccessTime;
+  private final boolean isNoContentSupported;
+
+  /**
+   * Constructs a request with a null access time that supports HTTP
+   * 204 responses. A null access times implies that the document has
+   * never been crawled.
+   *
+   * @param docId the requested document ID
+   */
+  public DocRequest(DocId docId) {
+    this(docId, null);
+  }
+
+  /**
+   * Constructs a request that supports HTTP 204 responses.
+   *
+   * @param docId the requested document ID
+   * @param lastAccessTime the last time the document was crawled
+   */
+  public DocRequest(DocId docId, Date lastAccessTime) {
+    this(docId, lastAccessTime, true);
+  }
+
+  /** Constructs a request.
+   *
+   * @param docId the requested document ID
+   * @param lastAccessTime the last time the document was crawled
+   * @param isNoContentSupported {@code true} if an HTTP 204 response
+   *     is allowed, and {@code false} otherwise
+   */
+  public DocRequest(DocId docId, Date lastAccessTime,
+      boolean isNoContentSupported) {
+    this.docId = docId;
+    this.lastAccessTime = lastAccessTime;
+    this.isNoContentSupported = isNoContentSupported;
+  }
```

Comment:
what does null doc id mean?

Reply:
GIGO. The library never passes in null, nor do our connectors, and this class does not (currently) deference it. I'll add a check in the constructor and throw a NPE.